src/Catel.MVVM/Collections/ObservableDictionary.cs

```diff
@@ -0,0 +1,682 @@
+            {
+                array[arrayIndex++] = new KeyValuePair<TKey, TValue>(keyValuePair.Key, keyValuePair.Value);
+            }
+        }
+
+        /// <inheritdoc cref="IEnumerable{T}.GetEnumerator"/>
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        /// <inheritdoc cref="IDictionary{TKey, TValue}.Remove(TKey)"/>
+        public bool Remove(TKey key)
+        {
+            if (!TryGetValue(key, out TValue value))
+            {
+                return false;
+            }
+
+            KeyValuePair<TKey, TValue> item = new KeyValuePair<TKey, TValue>(key, value);
+            int index = _collection.IndexOf(item);
+
+            if (!_collection.Remove(key))
+            {
+                return false;
+            }
+
+            OnCollectionChanged(NotifyCollectionChangedAction.Remove, item, index);
+
+            _version++;
+
+            return true;
+        }
+
+        /// <inheritdoc cref="IDictionary{TKey, TValue}.TryGetValue"/>
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            bool result = _collection.Contains(key);
+
+            value = result ? _collection[key].Value : default;
+
+            return result;
+        }
+
+        /// <inheritdoc cref="IEnumerable.GetEnumerator"/>
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        /// <inheritdoc cref="ICollection{T}.Remove"/>
+        bool ICollection<KeyValuePair<TKey, TValue>>.Remove(KeyValuePair<TKey, TValue> item)
+        {
+            return Remove(item.Key);
+        }
+
+        /// <inheritdoc cref="INotifyCollectionChanged.CollectionChanged"/>
+        public event NotifyCollectionChangedEventHandler CollectionChanged;
+
+        /// <inheritdoc cref="INotifyPropertyChanged.PropertyChanged"/>
+        public event PropertyChangedEventHandler PropertyChanged;
+
+        /// <inheritdoc cref="IReadOnlyDictionary{TKey, TValue}.Keys"/>
+        IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => Dictionary.Keys;
+
+        /// <inheritdoc cref="IReadOnlyDictionary{TKey, TValue}.Values"/>
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Dictionary.Values;
+
+        /// <inheritdoc cref="ISerializable.GetObjectData"/>
+        public virtual void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            Argument.IsNotNull(nameof(info), info);
+
+            Collection<KeyValuePair<TKey, TValue>> entries = new Collection<KeyValuePair<TKey, TValue>>();
+            foreach (KeyValuePair<TKey, TValue> item in _collection)
+            {
+                entries.Add(item);
+            }
+
+            info.AddValue(nameof(entries), entries);
+        }
+
+        private static bool IsCompatibleKey(object key)
+        {
+            Argument.IsNotNull(nameof(key), key);
```

Comment:
Maybe remove the check and do `return (!(key is null) && key is TKey)`

Reply:
This was actually simplified even further to be just `return key is TKey`